File: CS/WebSite/App_Code/ExportGridView.cs
using System;
using System.Data;
using System.Web;
using DevExpress.Web.ASPxGridView;
using DevExpress.Web.ASPxGridView.Export;
using DevExpress.Web.ASPxPager;
using System.Diagnostics;
using System.Web.UI;
using System.Collections.Generic;
using System.ComponentModel;

namespace Grid
{
    public class ExportGridView : DevExpress.Web.ASPxGridView.ASPxGridView
    {
        #region Fields (1)

        protected ExportType _ExportTypes = ExportType.none;

        #endregion Fields

        #region Constructors (1)

        public ExportGridView()
        {
            this.RegisterEvents();
        }

        #endregion Constructors

        #region Properties (6)

        public ExportType ExportTypes
        {
            get { return _ExportTypes; }
            set { _ExportTypes = value; }
        }

        public override string ID
        {
            get
            {
                return base.ID;
            }
            set
            {
                base.ID = value;
                this.ClientInstanceName = ID;
            }
        }

        protected byte[] SessionExportData
        {
            get
            {
                if (HttpContext.Current.Session[this.SessionExportDataKey] == null)
                    HttpContext.Current.Session[this.SessionExportDataKey] = new byte[] { };
                return (byte[])HttpContext.Current.Session[this.SessionExportDataKey];
            }
            set { HttpContext.Current.Session[this.SessionExportDataKey] = value; }
        }

        protected String SessionExportDataKey { get { return String.Format("{0}_ExportData", this.ID); } }

        protected ExportType SessionExportType
        {
            get
            {
                if (HttpContext.Current.Session[this.SessionExportTypeKey] == null)
                    HttpContext.Current.Session[this.SessionExportTypeKey] = ExportType.none;
                return (ExportType)HttpContext.Current.Session[this.SessionExportTypeKey];
            }
            set { HttpContext.Current.Session[this.SessionExportTypeKey] = value; }
        }

        protected String SessionExportTypeKey { get { return String.Format("{0}_ExportType", this.ID); } }

        #endregion Properties

        #region Methods (8)

        // Protected Methods (8) 

        protected void CheckExport()
        {
            ExportType type = this.SessionExportType;
            if (type != ExportType.none)
            {
                this.SessionExportType = ExportType.none;
                byte[] data = this.SessionExportData;
                this.SessionExportData = null;
                if (data != null)
                {
                    ExportHelper.WriteDataToResponse(data, type);
                }
            }
        }

        protected ASPxGridViewExporter GetGridExporter()
        {
            foreach (Control item in this.Controls)
            {
                ASPxGridViewExporter exp = item as ASPxGridViewExporter;
                if (exp != null)
                {
                    return exp;
                }
            }
            return null;
        }

        protected void LoadPager()
        {
            this.SettingsPager.AlwaysShowPager = true;
            ExportGridViewPager pager = new ExportGridViewPager();
            pager.ExportTypes = this.ExportTypes;
            this.Templates.PagerBar = pager;
        }

        protected override void OnDataBound(EventArgs e)
        {
            base.OnDataBound(e);
            this.CheckExport();
        }

        protected virtual void OnEventExportData(ExportType args)
        {
            this.DataBind();
            this.SessionExportType = args;
            this.SessionExportData = ExportHelper.GetExportData(this.GetGridExporter(), args);
            ASPxGridView.RedirectOnCallback(Request.Url.AbsolutePath);
        }

        protected void OnGridCustomCallback(object sender, ASPxGridViewCustomCallbackEventArgs e)
        {
            if (!String.IsNullOrEmpty(e.Parameters))
            {
                String[] values = e.Parameters.Split(new String[] { "||||" }, StringSplitOptions.RemoveEmptyEntries);
                Trace.WriteLine(String.Format("OnGridCustomCallBack - {0}", String.Join(", ", values)));
                if (values.Length == 3)
                {
                    switch (values[0])
                    {
                        case "EXPORT":
                            this.OnEventExportData((ExportType)Enum.Parse(typeof(ExportType), values[2]));
                            break;
                    }
                }
            }
        }

        protected override void OnInit(EventArgs e)
        {
            base.OnInit(e);
            this.LoadPager();
        }

        protected void RegisterEvents()
        {
            this.CustomCallback += this.OnGridCustomCallback;
        }

        #endregion Methods
    }
}

File: VB/WebSite/App_Code/ExportGridView.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Data
Imports System.Web
Imports DevExpress.Web.ASPxGridView
Imports DevExpress.Web.ASPxGridView.Export
Imports DevExpress.Web.ASPxPager
Imports System.Diagnostics
Imports System.Web.UI
Imports System.Collections.Generic
Imports System.ComponentModel

Namespace Grid
	Public Class ExportGridView
		Inherits DevExpress.Web.ASPxGridView.ASPxGridView

        Protected _ExportTypes As ExportType = ExportType.none

        Public Sub New()
            Me.RegisterEvents()
        End Sub

        Public Property ExportTypes() As ExportType
            Get
                Return _ExportTypes
            End Get
            Set(ByVal value As ExportType)
                _ExportTypes = value
            End Set
        End Property

        Public Overrides Property ID() As String
            Get
                Return MyBase.ID
            End Get
            Set(ByVal value As String)
                MyBase.ID = value
                Me.ClientInstanceName = ID
            End Set
        End Property

        Protected Property SessionExportData() As Byte()
            Get
                If HttpContext.Current.Session(Me.SessionExportDataKey) Is Nothing Then
                    HttpContext.Current.Session(Me.SessionExportDataKey) = New Byte() {}
                End If
                Return CType(HttpContext.Current.Session(Me.SessionExportDataKey), Byte())
            End Get
            Set(ByVal value As Byte())
                HttpContext.Current.Session(Me.SessionExportDataKey) = value
            End Set
        End Property

        Protected ReadOnly Property SessionExportDataKey() As String
            Get
                Return String.Format("{0}_ExportData", Me.ID)
            End Get
        End Property

        Protected Property SessionExportType() As ExportType
            Get
                If HttpContext.Current.Session(Me.SessionExportTypeKey) Is Nothing Then
                    HttpContext.Current.Session(Me.SessionExportTypeKey) = ExportType.none
                End If
                Return CType(HttpContext.Current.Session(Me.SessionExportTypeKey), ExportType)
            End Get
            Set(ByVal value As ExportType)
                HttpContext.Current.Session(Me.SessionExportTypeKey) = value
            End Set
        End Property

        Protected ReadOnly Property SessionExportTypeKey() As String
            Get
                Return String.Format("{0}_ExportType", Me.ID)
            End Get
        End Property

        ' Protected Methods (8) 

        Protected Sub CheckExport()
            Dim type As ExportType = Me.SessionExportType
            If type <> ExportType.none Then
                Me.SessionExportType = ExportType.none
                Dim data() As Byte = Me.SessionExportData
                Me.SessionExportData = Nothing
                If data IsNot Nothing Then
                    ExportHelper.WriteDataToResponse(data, type)
                End If
            End If
        End Sub

        Protected Function GetGridExporter() As ASPxGridViewExporter
            For Each item As Control In Me.Controls
                Dim exp As ASPxGridViewExporter = TryCast(item, ASPxGridViewExporter)
                If exp IsNot Nothing Then
                    Return exp
                End If
            Next item
            Return Nothing
        End Function

        Protected Sub LoadPager()
            Me.SettingsPager.AlwaysShowPager = True
            Dim pager As New ExportGridViewPager()
            pager.ExportTypes = Me.ExportTypes
            Me.Templates.PagerBar = pager
        End Sub

        Protected Overrides Sub OnDataBound(ByVal e As EventArgs)
            MyBase.OnDataBound(e)
            Me.CheckExport()
        End Sub

        Protected Overridable Sub OnEventExportData(ByVal args As ExportType)
            Me.DataBind()
            Me.SessionExportType = args
            Me.SessionExportData = ExportHelper.GetExportData(Me.GetGridExporter(), args)
            ASPxGridView.RedirectOnCallback(Request.Url.AbsolutePath)
        End Sub

        Protected Sub OnGridCustomCallback(ByVal sender As Object, ByVal e As ASPxGridViewCustomCallbackEventArgs)
            If (Not String.IsNullOrEmpty(e.Parameters)) Then
                Dim values() As String = e.Parameters.Split(New String() {"||||"}, StringSplitOptions.RemoveEmptyEntries)
                Trace.WriteLine(String.Format("OnGridCustomCallBack - {0}", String.Join(", ", values)))
                If values.Length = 3 Then
                    Select Case values(0)
                        Case "EXPORT"
                            Me.OnEventExportData(CType(System.Enum.Parse(GetType(ExportType), values(2)), ExportType))
                    End Select
                End If
            End If
        End Sub

        Protected Overrides Sub OnInit(ByVal e As EventArgs)
            MyBase.OnInit(e)
            Me.LoadPager()
        End Sub

        Protected Sub RegisterEvents()
            AddHandler Me.CustomCallback, AddressOf OnGridCustomCallback
        End Sub

    End Class
End Namespace
